arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -356,21 +343,26 @@ public fun <A> resource(block: suspend ResourceScope.() -> A): Resource<A> = blo
  * <!--- KNIT example-resource-06.kt -->
  */
 @ScopeDSL
-public suspend fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
-  val scope = ResourceScopeImpl()
-  val a: A = try {
-    action(scope)
+public suspend inline fun <A> resourceScope(action: ResourceScope.() -> A): A = with(ResourceScopeImpl()) {
+  try {
+    action()
   } catch (e: Throwable) {
-    val ex = if (e is CancellationException) ExitCase.Cancelled(e) else ExitCase.Failure(e)
-    val ee = withContext(NonCancellable) {
-      scope.cancelAll(ex, e) ?: e
-    }
-    throw ee
-  }
-  withContext(NonCancellable) {
-    scope.cancelAll(ExitCase.Completed)?.let { throw it }
-  }
-  return a
+    cancelAll(ExitCase(e))
+    error("Unreachable, cancelAll should throw the exception passed to it. Please report this bug to the Arrow team.")
+  }.also { cancelAll(Completed) }
+}
+
+/**
+ * Install [A] into the [ResourceScope].
+ * It's [release] function will be called with the appropriate [ExitCase] if this [ResourceScope] finishes.
+ * It results either in [ExitCase.Completed], [ExitCase.Cancelled] or [ExitCase.Failure] depending on the terminal state of [Resource] lambda.
+ */
+@ResourceDSL
+public suspend inline fun <A> ResourceScope.install(
```

Comment:
Why are these being defined as extension functions, this breaks source compatibility with 1.x.x. Why not define this in the interface directly?